### PR TITLE
Force cache flush if Redis went away

### LINKED ISSUE #4: Fall back to non-persistent object caching if Redis is unavailable
[27-Jun-2015 23:12:44 UTC] PHP Fatal error:  Uncaught exception 'RedisException' with message 'Redis server went away' in PATH/object-cache.php:713
[28-Jun-2015 03:10:20 UTC] PHP Fatal error:  Uncaught exception 'RedisException' with message 'read error on connection' in PATH/object-cache.php:652

There should be some handling in place for these errors. Ideally, we would be catching these exceptions and dumping them to the error log so that we know when these errors are occurring as well as what the errors are, but the site should simply fall back to connecting to the DB for its info, rather than fatal erring.

### PR DESCRIPTION
**Pull request by danielbachhuber**
WP Redis now falls back to its internal object cache if Redis goes away
during a request. However, as soon as WP Redis uses its internal object
cache, the Redis cache should be considered stale. As such, when we
successfully reconnect to Redis, we should force a Redis flush.

To keep track of whether Redis needs a forced flush, we need to write a
database option when it goes away, and make a direct SQL query to check
that database option on each load.

Users wary of the SQL reads / writes can define a
`WP_REDIS_DISABLE_WAKEUP_FLUSH` constant, which will mean Redis could
serve stale cache.

Fixes #4

### CODE REVIEW

**Comment by danielbachhuber** [comment]
@joshkoenig #reviewmerge

I feel good, but not great, about the test coverage on this, and look forward to your critique :)

File: object-cache.php [hunk 1]
```diff
@@ -762,6 +771,11 @@ protected function _call_redis( $method ) {
 			}
 		}
 
+		if ( $this->is_redis_wakeup_flush_enabled() && ! $this->redis_needs_flush ) {
```
Comment (joshkoenig) [inline object-cache.php]: Given than naming is one of the two hard problems, can I request we call this feature the `redis_failback_flush`?

The term fail-back is a generally used thing in HA systems, and it's widely accepted that "failing back" to the nominally primary system is as risky and complex as the initial fail-over. That's the frame of reference I'd like to promote in thinking about this feature.

Comment (danielbachhuber) [inline object-cache.php]: > Given than naming is one of the two hard problems, can I request we call this feature the redis_failback_flush?

Yep, 9173188

**Comment by danielbachhuber** [comment]
Huh. Tests pass locally. Not sure why this error is triggered in 91731884f3f0058615f8d014e9ce66ae9aa10f3a but not c292d38c60c42aed4ab4c34b593bfb6f2b12e4ec when all I changed was the name of some variables.

**Comment by danielbachhuber** [comment]
Aha! Restarted the second build, and now it fails too. And I'm getting this error:

```
WordPress database error Table 'wordpress_test.wptests_options' doesn't exist for query SELECT option_value FROM wptests_options WHERE option_name='wp_redis_do_redis_fallback_flush' made by require_once('wp-settings.php'), wp_start_object_cache, wp_cache_init, WP_Object_Cache->__construct
<div id="error"><p class="wpdberror"><strong>WordPress database error:</strong> [Table &#039;wordpress_test.wptests_options&#039; doesn&#039;t exist]<br /><code>SELECT option_value FROM wptests_options WHERE option_name=&#039;wp_redis_do_redis_fallback_flush&#039;</code></p></div
```

**Comment by danielbachhuber** [comment]
Crazy. When I [branch from this branch](https://github.com/pantheon-systems/wp-redis/compare/4-flush-cache...pantheon-systems:debug-failure-2) and run the build, it passes: https://travis-ci.org/pantheon-systems/wp-redis/builds/89713693

**Comment by danielbachhuber** [comment]
I also tried [catching the Exception](https://github.com/pantheon-systems/wp-redis/commit/8a3f15bb3c12c9b3de684bf824db7267bfc64f94), but that didn't yield any results https://travis-ci.org/pantheon-systems/wp-redis/builds/89712663

**Comment by danielbachhuber** [comment]
Fixed! And boy, did this one destroy me.

**Comment by joshkoenig** [comment]
Thanks for the naming nitpick

**Pull request merged**